subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -16,102 +16,115 @@
-            Class<T> valueType = Cast.<ProviderInternal<T>>uncheckedCast(provider).getType();
-            if (valueType != null) {
-                assertAttributeTypeIsValid(valueType, key);
-            }
+
+        ProviderInternal<T> providerInternal = Cast.uncheckedCast(provider);
+
+        MappingProvider<Isolatable<T>, T> isolated;
+        Class<T> valueType = providerInternal.getType();
+        if (valueType != null) {
+            // We can only sometimes check the type of the provider ahead of time.
+            assertAttributeTypeIsValid(valueType, key);
+            isolated = new MappingProvider<>(null, providerInternal, attributesFactory::isolate);
+        } else {
+            // Otherwise, check the type when the value is realized.
+            isolated = new MappingProvider<>(null, providerInternal, t -> {
+                assertAttributeTypeIsValid(t.getClass(), key);
+                return attributesFactory.isolate(t);
+            });
         }
-        doInsertionLazy(key, provider);
+
+        state.put(key, isolated);
+
         return this;
     }
 
     private <T> void checkInsertionAllowed(Attribute<T> key) {
+        // TODO: This check should be handled by the provider API infrastructure
         if (realizingAttributes) {
             throw new IllegalStateException("Cannot add new attribute '" + key.getName() + "' while realizing all attributes of the container.");
         }
-        for (Attribute<?> attribute : keySet()) {
-            String name = key.getName();
-            if (attribute.getName().equals(name) && attribute.getType() != key.getType()) {
-                throw new IllegalArgumentException("Cannot have two attributes with the same name but different types. "
-                    + "This container already has an attribute named '" + name + "' of type '" + attribute.getType().getName()
-                    + "' and you are trying to store another one of type '" + key.getType().getName() + "'");
+    }
+
+    private Map<Attribute<?>, Isolatable<?>> getRealizedMap() {
+        Map<Attribute<?>, Isolatable<?>> realizedMap = doRealizeAttributes();
+
+        Map<String, Attribute<?>> attributesByName = new HashMap<>();
+        for (Map.Entry<Attribute<?>, Isolatable<?>> entry : realizedMap.entrySet()) {
+            Attribute<?> attribute = entry.getKey();
+            String name = attribute.getName();
+            Attribute<?> existing = attributesByName.put(name, attribute);
+            if (existing != null) {
+                throw new IllegalStateException("Cannot have two attributes with the same name but different types. "
+                    + "This container has an attribute named '" + name + "' of type '" + existing.getType().getName()
+                    + "' and another attribute of type '" + attribute.getType().getName() + "'");
             }
         }
+
+        return realizedMap;
+    }
+
+    private Map<Attribute<?>, Isolatable<?>> doRealizeAttributes() {
```

Comment:
🤔 It seems like this is trying to keep the time when insertion is not allowed to an absolute minimum.

But it seems weird to me that if we have a sequence like:

```
  asImmutable() called
    getRealizedMap() called
      getRealizedAttributes() is called...completes
  attribute 2 is inserted
    getRealizedMap() continues...loops...completes
  asImmutable() completes and returns
```

Then the result of `asImmutable` would **not** contain attribute 2, which was added prior to the completion of `getRealizedMap()`.  

Maybe it's better to be a little more pessimistic and lock insertion for the entirety of `getRealizedMap` instead (which would remove the need for `getRealizedAttributes()` entirely, as it could then be inlined.

Reply:
An `AttributeContainer` is _not_ thread safe -- the backing `MapProperty` as well is not thread safe. 

We do not have to worry about this situation

Reply:
Discussed and resolved.